masonry/src/passes/update.rs

```diff
@@ -0,0 +1,220 @@
+        .widgets
+        .find_mut(id.to_raw())
+        .expect("widget not found in arena");
+
+    // Box<dyn Widget> -> &dyn Widget
+    // Without this step, the type of `WidgetRef::widget` would be
+    // `&Box<dyn Widget> as &dyn Widget`, which would be an additional layer
+    // of indirection.
+    let widget = widget_mut.item;
+    let widget: &mut dyn Widget = &mut **widget;
+
+    (
+        widget,
+        PassCtx {
+            widget_state: state_mut,
+            widget_children: widget_mut.children,
+        },
+    )
+}
+
+// TODO - Replace LifecycleCtx with UpdateCtx
+pub(crate) fn run_targeted_update_pass(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    target: Option<WidgetId>,
+    pass_fn: impl FnMut(&mut dyn Widget, &mut LifeCycleCtx),
+) {
+    let mut pass_fn = pass_fn;
+
+    let mut target_widget_id = target;
+    while let Some(widget_id) = target_widget_id {
+        let (widget, pass_ctx) = get_widget_mut(&mut root.widget_arena, widget_id);
+        let parent_id = pass_ctx.parent();
+
+        let mut ctx = LifeCycleCtx {
+            global_state: &mut root.state,
+            widget_state: pass_ctx.widget_state.item,
+            widget_state_children: pass_ctx.widget_state.children,
+            widget_children: pass_ctx.widget_children,
+        };
+        pass_fn(widget, &mut ctx);
+
+        merge_state_up(&mut root.widget_arena, widget_id, root_state);
+        target_widget_id = parent_id;
+    }
+}
+
+pub(crate) fn run_update_pointer_pass(root: &mut RenderRoot, root_state: &mut WidgetState) {
+    let pointer_pos = root.last_mouse_pos.map(|pos| (pos.x, pos.y).into());
+
+    // -- UPDATE HOVERED WIDGETS --
+
+    let hovered_widget = root.state.hovered_path.last().copied();
+    let mut next_hovered_widget = if let Some(pos) = pointer_pos {
+        // TODO - Apply scale?
+        root.get_root_widget()
+            .find_widget_at_pos(pos)
+            .map(|widget| widget.id())
+    } else {
+        None
+    };
+    if let Some(capture_target) = root.state.pointer_capture_target {
+        if next_hovered_widget != Some(capture_target) {
+            next_hovered_widget = None;
+        }
+    }
+
+    let prev_hovered_path = std::mem::take(&mut root.state.hovered_path);
+    let next_hovered_path = get_id_path(root, next_hovered_widget);
+
+    let mut hovered_set = HashSet::new();
+    for widget_id in &next_hovered_path {
+        hovered_set.insert(*widget_id);
+    }
+
+    trace!("prev_hovered_path: {:?}", prev_hovered_path);
+    trace!("next_hovered_path: {:?}", next_hovered_path);
+
+    // This algorithm is written to be resilient to future changes like reparenting and multiple
+    // cursors. In theory it's O(Depth² * CursorCount) in the worst case, which isn't too bad
+    // (cursor count is usually 1 or 2, depth is usually small), but in practice it's almost
+    // always O(Depth * CursorCount) because we only need to update the hovered status of the
```

Comment:
Isn't that then more like the average case?

Reply:
It's not average or amortized, because by default after you visit a widget you immediately visit the next one in `next_hovered_path` (its parent).

To be specific, worst case for the algorithm below that comment is `O(Depth * CursorCount)` *unless* a widget has been reparented (which currently isn't possible but might be in the future) in which case it's `O(Depth² * CursorCount)`.

(The above assumes that accessing a widget is `O(1)`, which isn't the case but will be in the future.)

Reply:
Ok, so the factor of $$\text{Depth}^{2}$$ occurs if at every item in the tree, the parent has changed. I.e. the tree used to be:
```mermaid
graph TB;
    A;
    B;
    A-->C;
    A-->D;
    C-->E;
```
where E was hovered over, and is now something like:
```mermaid
graph TB;
    A;
    A-->D;
    B-->C;
    D-->E;
```
even if E is still hovered over?

Reply:
Yup.